proof/refine/ARM/Ipc_R.thy

```diff
@@ -3560,160 +3560,622 @@ lemma cteInsert_ct'[wp]:
+  apply (simp only: replyPush_def)
+  apply (rule replyPush_sym_refs_helper hoare_fun_app_wp hoare_when_wp hoare_seq_ext hoare_K_bind)+
+                      apply (wpsimp wp: threadGet_wp hoare_vcg_all_lift hoare_vcg_if_lift
+                                        hoare_vcg_imp_lift' hoare_vcg_disj_lift gts_wp'
+                                        setReply_obj_at' haskell_assert_wp
+                                  simp: setReplyTCB_def getReplyTCB_def)+
+  apply (fastforce simp: comp_def obj_at'_def projectKO_eq projectKO_tcb)
+  done
+
+lemma replyPush_if_live_then_nonz_cap':
+  "\<lbrace>\<lambda>s. if_live_then_nonz_cap' s \<and> valid_objs' s \<and> valid_idle' s \<and> st_tcb_at' active' callerPtr s \<and>
+        ex_nonz_cap_to' replyPtr s \<and> ex_nonz_cap_to' callerPtr s \<and> ex_nonz_cap_to' calleePtr s\<rbrace>
+   replyPush callerPtr calleePtr replyPtr canDonate
+   \<lbrace>\<lambda>_. if_live_then_nonz_cap'\<rbrace>"
+  apply (rule monadic_rewrite_refine_valid[OF monadic_rewrite_replyPush, where P''=\<top>, simplified])
+  apply (wp schedContextDonate_if_live_then_nonz_cap')
+        apply (rule_tac Q="\<lambda>_ s. (\<forall>scp. scPtrOptDonated = Some scp \<longrightarrow> ex_nonz_cap_to' scp s) \<and>
+                                 ex_nonz_cap_to' calleePtr s \<and>
+                                 valid_objs' s \<and> if_live_then_nonz_cap' s"
+                     in hoare_strengthen_post[rotated], clarsimp)
+        apply (wp hoare_vcg_all_lift hoare_vcg_const_imp_lift)
+        apply (simp (no_asm_use) add: replyPush'_def getReplyTCB_def split del: if_split
+               | wp hoare_vcg_all_lift hoare_vcg_disj_lift hoare_vcg_imp_lift'
+                    hoare_vcg_if_lift setReply_obj_at' setReplyTCB_obj_at'
+               | rule threadGet_wp)+
+  apply (clarsimp simp: obj_at'_def projectKO_eq projectKO_tcb projectKO_sc projectKO_reply)
+  apply (rule context_conjI; clarsimp simp: sym_refs_asrt_def)
+   apply (erule if_live_then_nonz_capE')
+   apply (drule_tac ko=obj in sym_refs_ko_atD'[rotated])
+    apply (auto simp: refs_of_rev' live_sc'_def idle_tcb'_def valid_idle'_def
+                      pred_tcb_at'_def ko_wp_at'_def obj_at'_def projectKO_eq projectKO_tcb)[2]
+  apply (fastforce elim: if_live_then_nonz_capE' dest: sym_refs_ko_atD'[rotated]
+                   simp: ko_wp_at'_def obj_at'_def projectKO_eq refs_of_rev' live_reply'_def)
+  done
+
+lemma replyPush_valid_idle':
+  "\<lbrace>valid_idle' and valid_pspace' and st_tcb_at' active' callerPtr\<rbrace>
+   replyPush callerPtr calleePtr replyPtr canDonate
+   \<lbrace>\<lambda>_. valid_idle'\<rbrace>"
+  apply (simp only: replyPush_def)
+  apply (simp split del: if_split
+         | wp hoare_vcg_if_lift[where f="setReply _ _"] hoare_vcg_imp_lift'
+              hoare_vcg_all_lift setReply_obj_at' schedContextDonate_valid_idle'
+         | (rule hoare_lift_Pf3[where f=ksIdleThread,rotated]; wp setSchedContext_obj_at'))+
```

Comment:
I'm fine with this if it's necessary, but this is a bit of a messy apply step. It definitely doesn't work with some combination of `wpsimp` and `wps`?

Reply:
> I'm fine with this if it's necessary, but this is a bit of a messy apply step. It definitely doesn't work with some combination of `wpsimp` and `wps`?

This is about as simple as I can make it, unfortunately. The simp step needs to take precedence over the wp step, and the application of hoare_lift_Pf3 causes looping when I use it with the other wp rules.